bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch

```diff
@@ -0,0 +1,3422 @@
++    do {
++      const double total_tokens = std::min(max_tokens_, (time_now - time_old) * fill_rate_);
++      if (consumed = cb(total_tokens); consumed == 0) {
++        return 0;
++      }
++
++      // There are two special cases that should rarely happen in practice but we will not
++      // prevent them in this common template method:
++      // The consumed is negative. It means the token is added back to the bucket.
++      // The consumed is larger than total_tokens. It means the bucket is overflowed and future
++      // tokens are consumed.
++
++      // Move the time_in_seconds_ forward by the number of tokens consumed.
++      const double total_tokens_new = total_tokens - consumed;
++      time_new = time_now - (total_tokens_new / fill_rate_);
++    } while (
++        !time_in_seconds_.compare_exchange_weak(time_old, time_new, std::memory_order_relaxed));
++
++    return consumed;
++  }
++
++  /**
++   * Consumes one tokens from the bucket.
++   * @return true if the token is consumed, false otherwise.
++   */
++  bool consume();
++
++  /**
++   * Consumes multiple tokens from the bucket.
++   * @param tokens the number of tokens to consume.
++   * @param allow_partial whether to allow partial consumption.
++   * @return the number of tokens consumed.
++   */
++  uint64_t consume(uint64_t tokens, bool allow_partial);
++
++  /**
++   * Get the maximum number of tokens in the bucket. The actual maximum number of tokens in the
++   * bucket may be changed with the factor.
++   * @return the maximum number of tokens in the bucket.
++   */
++  double maxTokens() const { return max_tokens_; }
++
++  /**
++   * Get the fill rate of the bucket. This is a constant for the lifetime of the bucket. But note
++   * the actual used fill rate will multiply the dynamic factor.
++   * @return the fill rate of the bucket.
++   */
++  double fillRate() const { return fill_rate_; }
++
++  /**
++   * Get the remaining number of tokens in the bucket. This is a snapshot and may change after the
++   * call.
++   * @return the remaining number of tokens in the bucket.
++   */
++  double remainingTokens() const;
++
++private:
++  double timeNowInSeconds() const;
++
++  const double max_tokens_;
++  const double fill_rate_;
++
++  std::atomic<double> time_in_seconds_{};
++  TimeSource& time_source_;
++};
++
+ } // namespace Envoy
+diff --git a/source/common/runtime/runtime_features.cc b/source/common/runtime/runtime_features.cc
+index 260a08820c..f22d9cf58b 100644
+--- source/common/runtime/runtime_features.cc
++++ source/common/runtime/runtime_features.cc
+@@ -73,6 +73,7 @@ RUNTIME_GUARD(envoy_reloadable_features_lowercase_scheme);
+ RUNTIME_GUARD(envoy_reloadable_features_no_downgrade_to_canonical_name);
+ RUNTIME_GUARD(envoy_reloadable_features_no_extension_lookup_by_name);
+ RUNTIME_GUARD(envoy_reloadable_features_no_full_scan_certs_on_sni_mismatch);
++// RUNTIME_GUARD(envoy_reloadable_features_no_timer_based_rate_limit_token_bucket);
```

Comment:
TODO remove this

Reply:
I explicitly want this left in as it is what the actual upstream version will have.